### ISSUE TITLE
Performance

### ISSUE THREAD

**Issue by SafelySwift** [wren-lang/wren#653]
> Likewise, when you access a field in other languages, the interpreter has to look it up by name in a hash table in the object.

I ran some tests in Swift:
- 0.0005 for hash table
- 0.00000618 for pointers

Sorry for the beginner question, but why does the hash table approach differ by **orders of magnitude** in performance? I did not expect this.

**Comment by mhermier** [comment]
Because most of the time an object can't store its hash. And even when a
class is wrote most of the time developers forgot or can't store the hash
for various reasons. So instead of a linear memory access, you have a
triple penalty when trying to access a hash. You have to compute the hash
of you key, you get somehow a possible linear index position. With it you
have to do possibly multiple comparison in a linear search, meaning you
have to do a linear access to verify key equality. So even if by chance
there is no key collision in the hash table, getting your value is at
minimum: 1 hash computation, 1 equality comparison and 2 linear access to
get the stored key and value. Compared to a pointer (linear) access, you
get your order of magnitude.

And this is with our implementation. Hash table are meant to be used with
huge data set. And our *naive* implementation is a simple solution. There
are implementation in the wild with even more magnitude because of
serialization to discs because the data would not fit in memory.

In conclusion hashset are not free at all, so you have to avoid them when
possible, but they are unbeatable when you need to handle large data set.

>

**Comment by SafelySwift** [comment]
I implemented this object system in Swift and the fastest time was 0.0001 seconds. C++ took orders of magnitude less than this. Why the difference because Swift took the same time as C++.

**Comment by mhermier** [comment]
Then I don't understand your question. You ask why it is slow in wren? I
don't have a clue. It should be quite fast still. Do you have an example to
share? Maybe your example/test is *broken*

**Comment by SafelySwift** [comment]
I asked why it is so much slower implementing objects in Swift than it is actually using the in C++. 

Here is my Swift code:

```
//
//  Object.swift
//  Swizzle - Compiling
//
//  Created by Ethan Uppal on 1/16/19.
//  Copyright © 2019 Ethan Uppal. All rights reserved.
//

import Foundation

public struct InitializationError: Error {
    public let message: String
    public let context: Any?
}

public struct TypeError: LocalizedError {
    public let given: String
    public let expected: String
    public var errorDescription: String? {
        return "Cannot convert value of type \(given) ro type \(expected)."
    }
}

public enum SwizzleValue {
    case object(SwizzleObject)
    case string(String)
    case int(Int)
    case float(Float)
    case bool(Bool)
    public var type: String {
        switch self {
        case .object(let obj):
            return obj.type
        case .string: return "String"
        case .int: return "Int"
        case .float: return "Float"
        case .bool: return "Bool"
        default:
            fatalError("UNknown type for case in SwizzleValue.type")
        }
    }
}

extension SwizzleValue: ExpressibleByIntegerLiteral, ExpressibleByFloatLiteral, ExpressibleByStringLiteral, ExpressibleByBooleanLiteral {
    public init(integerLiteral value: Int) {
        self = .int(value)
    }
    public init(floatLiteral value: Float) {
        self = .float(value)
    }
    public init(stringLiteral value: String) {
        self = .string(value)
    }
    public init(booleanLiteral value: Bool) {
        self = .bool(value)
    }
}

public final class SwizzleObject {
    public let type: String
    var storage: UnsafeMutablePointer<SwizzleValue>
    let size: Int
    public init(source: ObjectStatement, parameters: [SwizzleValue]) throws {
        self.size = parameters.count
        self.type = source.name.lexme
        guard size == source.references.count else {
            throw InitializationError(message: "Expected \(size) arguments in initializer for \(type).", context: nil)
        }
        var ps = CFAbsoluteTimeGetCurrent()
        self.storage = UnsafeMutablePointer.allocate(capacity: size)
        print("ptr:", CFAbsoluteTimeGetCurrent() - ps)
        for index in 0 ..< size {
            let param = parameters[index]
            storage[index] = param
            let given = source.references[index].type.lexme
            let expected = param.type
            guard given == expected else {
                throw TypeError(given: given, expected: expected)
            }
        }
    }
    
    public func lookup(index: Int) -> SwizzleValue {
        return storage[index]
    }
    public func modify(index: Int, value: SwizzleValue) {
        storage[index] = value
    }
    
    deinit {
        storage.deallocate()
    }
}
```

**Comment by munificent** [comment]
I'm a little confused. Does this question have anything to do with Wren?

**Comment by SafelySwift** [comment]
My question is, how does wren implement objects so that there is no performance loss. This is stated in the documentation.

**Comment by mhermier** [comment]
Well I would say that contrary to most other dynamic language, object/class
are not made using a hashtable, but with a plain old data structure. So
objects have a fixed layout, which is easy to access if you compare to
dynamic objects like in js/Lua.
But I still wonder if this is really the explanation you are expecting.

**Comment by SafelySwift** [comment]
My question is: how do you lookup a string named property if the Noel is a pointer. What optimizations do you make to speed up the object? Detailed explanation.

**Comment by mhermier** [comment]
Wren String is immutable, as a consequence its hash is computed at
construction and cached. Second optimisation, map is a native hashmap with
a lot of restrictions on the key types. These restrictions are basically:
only allow keys that are VM built-in. So the default map basically works
with C code at *full speed*. The drawback is that you can't define your own
key type.

**Comment by SafelySwift** [comment]
That is not my question. My question is:

When you parse, you get the following tokens "foo", ".", "bar". During compilation you push that value on to the stack. How do you get the value of this statement if all you have is a pointer, since pointes cannot be accessed with "bar", buy only with integers.

**Comment by mhermier** [comment]
In case of wren "foo" "." "bar" express a method call.
And I think you are confused. A pointer is not the grand truth about
computer. Compilers can hide you some complexity. And evaluating an
expression doesn't always mean it is evaluate the same way in every
language.

>

**Comment by mhermier** [comment]
Compiler knows the correspondence. Like in C or any language that
translates members veritable to an offset.
The name is just a fancy way to write stuff for humans. Compiler translate
that to a pointer or an offset or whatever you need to retrieve it.

>

**Comment by underscorediscovery** [comment]
I think the question is maybe about the lookup part? When the wren compiler is going through emitting bytecode, it makes a symbol table that stores simple integer indexes into an array of `Value`s. Methods also store args then locals in a similar way to find their value. 

`var some = other.thing` is a few things. If you break it down there is:
- `some` is a local symbol, it is stored at index ~`args count+0`
- other is a symbol, looked up from the symbol table, `Value other_value = symbols[other_index]`
- `.thing` is a method call. it does not look into the object other as a hash or object. it just calls a method, which has it's own locals+args, and repeats the process.
- assign result of method call to value at local symbol `some`
 
Hopefully since I'm in a hurry I didn't misremember but hope that adds some useful details.

**Comment by SafelySwift** [comment]
> symbol table that stores simple integer indexes

This is a hash table. I though you said you did not use a hash table. I was wondering then how you get an integer index from a method name.

**Comment by underscorediscovery** [comment]
hmm, though an array isn't a hash table... the symbols are indexes in an array. There is no lookup, only array access.

**Comment by SafelySwift** [comment]
But a method is (tokenized to) a string. You said that values are stored in an array. How can you index an array with a string?

**Comment by underscorediscovery** [comment]
There are a few timeframes here - source, compile time, and runtime. Before compile time, the method is a string, but during parsing/compilation (which is one pass in wren), the string is parsed into a token (lexing/parsing) and then _compiled_ into a symbol table. 

**compile time** 

The below is from `wren_compiler.c`

```c++
static int declareVariable(Compiler* compiler, Token* token) {
....
    int symbol = wrenDefineVariable(compiler->parser->vm,
                                    compiler->parser->module,
                                    token->start, token->length, NULL_VAL);
```

Notice the return value, it's an int. The int is the array index. The symbol table is an array. This has nothing to do with runtime yet. Now that the compiler has a symbol, it is emitted into the bytecode. Here's an example from the source.

```c++
// Stores a variable with the previously defined symbol in the current scope.
static void defineVariable(Compiler* compiler, int symbol)
{
  // Store the variable. If it's a local, the result of the initializer is
  // in the correct slot on the stack already so we're done.
  if (compiler->scopeDepth >= 0) return;

  // It's a module-level variable, so store the value in the module slot and
  // then discard the temporary for the initializer.
  emitShortArg(compiler, CODE_STORE_MODULE_VAR, symbol);
  emitOp(compiler, CODE_POP);
}
```

**runtime**

Then the bytecode is used at runtime by the virtual machine. This is in `wren_vm.c`. 
The runtime uses the generated symbol table that the compiler built to operate. This is an array.

Here's an example from the vm. It reads a field member.
`uint8_t` is the symbol of the field, it's the array index into the `fields` symbol table of the class. `instance->fields[field]` is the array read. 

```c++
    CASE_CODE(LOAD_FIELD):
    {
      uint8_t field = READ_BYTE();
      Value receiver = POP();
      ASSERT(IS_INSTANCE(receiver), "Receiver should be instance.");
      ObjInstance* instance = AS_INSTANCE(receiver);
      ASSERT(field < instance->obj.classObj->numFields, "Out of bounds field.");
      PUSH(instance->fields[field]);
      DISPATCH();
    }
```

I suspect that the above gives you the tools to investigate the source in wren and answer the questions you might have on your own. Wren is lovingly commented for this intention so I hope you can find the insight you're looking for there, a lot of the stuff you're asking is answered with a little reading of the code :) 

While this thread has been pretty interesting and the discussion can continue still, I'm gonna close it for now as it is just a discussion.

**Comment by SafelySwift** [comment]
Sorry, but I am not very advanced in C so can you explain to a spoiled Swift developer what `wrenDefineVariable` does to convert the token into an integer?

**Comment by underscorediscovery** [comment]
It does:
1. checks if there's too many variables declared
2. if it's an object (not a primitive) stores it in a GC root, so it doesn't get collected while operating on it
3. searches in the symbol table array to see if the variable is already declared. this is just an array so it just does a linear search. Since it's not a high number (<hundreds) this is simpler memory access pattern and is fast.
4.  if not found, add it to the array, return the index
5. if it was implicitly defined, update the index to match the actual value
6. remove from the GC roots if 2 happened
7. return the array index

I also wanted to add a note about the memory thing, which you may or may not know but for completeness sake:

A lot of what makes code fast these days on modern hardware has a lot to do with the [memory access patterns](https://imgur.com/kOyKsSm). When memory access misses the CPU caches, it takes an heck of a lot more time (relative) to operate. Data structures that have good memory access like arrays are always faster than ones like hashes that have less coherence, specifically because the layout in memory is what the CPU is good at doing. 

Writing fast code is all about writing code with good access patterns, trying to get as much of the data pipelined into the CPU so the actual code instructions are never waiting for data to load, it's always there (it gets prefetched). A lot of the decisions in wren (like linear search through an array) are trade offs around this principle, because it will be faster than waiting for a cache miss.

Here's [a chapter from munificent's game programming book about data locality](http://gameprogrammingpatterns.com/data-locality.html) with some more details, and 'data oriented design' is the common term for this method of treating the data layout as most important.

**Comment by bjorn** [comment]
It does the most basic thing you can imagine:

* It does a linear search in the current symbols array, to see if there is already a symbol with that name defined.
* If it finds one, returns -1 (this will cause the "Module variable is already defined." for example).
* If it doesn't find out, appends a new entry to the array and returns array length - 1.

The point remains, that at compile-time, there will need to be string lookups (be it searching through arrays or hash maps), but at runtime it can simply use the index in an array to get/set a variable or field.

Source links:

https://github.com/wren-lang/wren/blob/f23c82071a84f38a51189e115ca409f5f57cad53/src/vm/wren_vm.c#L1489
https://github.com/wren-lang/wren/blob/40c927f4402bb6ff74fe8aa257bf8042eeff6544/src/vm/wren_utils.c#L43
https://github.com/wren-lang/wren/blob/40c927f4402bb6ff74fe8aa257bf8042eeff6544/src/vm/wren_utils.c#L20

**Comment by SafelySwift** [comment]
Oh, so you have 1 pointer for all the objects?

**Comment by SafelySwift** [comment]
So you have an array of boxed types. When you call a method or such you access the value, and then cast it to the necessary type for wren's dynamic-ness?
